core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -62,14 +64,70 @@ object NonEmptyLazyList extends NonEmptyLazyListInstances {
     if (as.nonEmpty) Option(create(LazyList.from(as))) else None
 
-  def fromLazyListPrepend[A](a: A, ca: LazyList[A]): NonEmptyLazyList[A] =
-    create(a +: ca)
+  def fromLazyListPrepend[A](a: => A, ll: => LazyList[A]): NonEmptyLazyList[A] =
+    create(a #:: ll)
+
+  @deprecated("Use overload with by-name args", "2.11.0")
+  def fromLazyListPrepend[A]()(a: A, ll: LazyList[A]): NonEmptyLazyList[A] =
+    fromLazyListPrepend(a, ll)
+
+  def fromLazyListAppend[A](ll: => LazyList[A], a: => A): NonEmptyLazyList[A] =
+    create(ll #::: a #:: LazyList.empty)
 
-  def fromLazyListAppend[A](ca: LazyList[A], a: A): NonEmptyLazyList[A] =
-    create(ca :+ a)
+  @deprecated("Use overload with by-name args", "2.11.0")
+  def fromLazyListAppend[A]()(ll: LazyList[A], a: A): NonEmptyLazyList[A] =
+    fromLazyListAppend(ll, a)
 
   def apply[A](a: => A, as: A*): NonEmptyLazyList[A] =
-    create(LazyList.concat(LazyList(a), LazyList.from(as)))
+    create(a #:: LazyList.from(as))
+
+  /**
+   * Wraps a `LazyList` that may or may not be empty so that individual
+   * elements or `NonEmptyLazyList`s can be prepended to it to construct a
+   * result that is guaranteed not to be empty without evaluating any elements.
+   */
+  def maybe[A](ll: => LazyList[A]): Maybe[A] = new Maybe(() => ll)
+
+  /**
+   * Wrapper around a `LazyList` that may or may not be empty so that individual
+   * elements or `NonEmptyLazyList`s can be prepended to it to construct a
+   * result that is guaranteed not to be empty without evaluating any elements.
+   */
+  final class Maybe[A] private[NonEmptyLazyList] (mkLL: () => LazyList[A]) {
+    // because instances of this class are created explicitly, they might be
+    // reused, and we don't want to re-evaluate `mkLL`
+    private[this] lazy val ll = mkLL()
+
+    /** Prepends a single element, yielding a `NonEmptyLazyList`. */
+    def #::[AA >: A](elem: => AA): NonEmptyLazyList[AA] =
+      create(elem #:: ll)
+
+    /** Prepends a `LazyList`, yielding another [[Maybe]]. */
+    def #:::[AA >: A](prefix: => LazyList[AA]): Maybe[AA] =
+      new Maybe(() => prefix #::: ll)
+
+    /** Prepends a `NonEmptyLazyList`, yielding a `NonEmptyLazyList`. */
+    def #:::[AA >: A](prefix: => NonEmptyLazyList[AA])(implicit d: DummyImplicit): NonEmptyLazyList[AA] =
+      create(prefix.toLazyList #::: ll)
+  }
+
+  final class Deferrer[A] private[NonEmptyLazyList] (private val nell: () => NonEmptyLazyList[A]) extends AnyVal {
+
+    /** Prepends a single element. */
+    def #::[AA >: A](elem: => AA): NonEmptyLazyList[AA] =
+      create(elem #:: nell().toLazyList)
+
+    /** Prepends a `LazyList`. */
+    def #:::[AA >: A](prefix: => LazyList[AA]): NonEmptyLazyList[AA] =
+      create(prefix #::: nell().toLazyList)
+
+    /** Prepends a `NonEmptyLazyList`. */
+    def #:::[AA >: A](prefix: => NonEmptyLazyList[AA])(implicit d: DummyImplicit): NonEmptyLazyList[AA] =
+      create(prefix.toLazyList #::: nell().toLazyList)
```

Comment:
Do we really need this overload with `NonEmptyLazyList` type if we already have another one that takes `LazyList`? I am a bit concerned about this overload trick. Usually, overloaded method names are discouraged unless absolutely necessary. But it does not seem to be the case here, because prepending a `NonEmptyLazyList` can be easily achieved with just the single method above that takes `LazyList` only:
```scala
def a: NonEmptyLasyList[X] = ???
def b: NonEmptyLasyList[X] = ???

val c: NonEmptyLazyList[X] = a.toLazyList #::: b
```

Reply:
it would be weird to have `NonEmptyLazyList` be a second-class citizen of its own API. however, the meaning of the methods are identical, yielding no additional conceptual weight and making the API easier to use. additionally, it's not possible to accidentally use the wrong method either, as they are only the same after erasure. I don't personally see any issue with it

Reply:
But it pretty much corresponds to the way how other APIs are organized in Cats. See for example `NonEmptyList.concat` vs `NonEmptyList.concatNel`, also `NonEmptyVector.concat` vs `concatNev`, etc. Retrospectively, it would not be a problem to overload `concat` everywhere in the way as you're proposing. But for some reason the decision was to avoid overloading and provide clearer separate names wherever possible. I personally support it because feel a bit sour about having any sort of "hacks and tricks" in code – they are usually easy to add but tend to turn into maintenance hell in long-term run.

That said, `NonEmptyList` has the `:::` which takes another `NonEmptyList`, not just `List`. Perhaps it is just a legacy of that particular class because it was the very first `NonEmpty*` one. Anyway, it is still not overloaded.